{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ------- SERVER EXTENSIONS ---------\n",
    "lib =  r'/srv01/technion/shitay/Code/classifying_response_to_immunotherapy/utilities/droplet_dataset'\n",
    "lib2 = r'/srv01/technion/shitay/Code/classifying_response_to_immunotherapy/utilities'\n",
    "lib3 = r'/srv01/technion/shitay/Code/classifying_response_to_immunotherapy/data_analysis'\n",
    "lib4 = r'/srv01/technion/shitay/Code/classifying_response_to_immunotherapy'\n",
    "lib5 = r'/srv01/technion/shitay/Code/classifying_response_to_immunotherapy/scripts'\n",
    "import sys\n",
    "sys.path.append(lib)\n",
    "sys.path.append(lib2)\n",
    "sys.path.append(lib3)\n",
    "sys.path.append(lib4)\n",
    "sys.path.append(lib5)\n",
    "# ------- SERVER EXTENSIONS ---------\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy\n",
    "import sklearn\n",
    "from sklearn.manifold import TSNE\n",
    "import pickle\n",
    "# from Bio.Cluster import kcluster\n",
    "import os\n",
    "import numpy as np\n",
    "import yaml\n",
    "import os\n",
    "import math\n",
    "import pandas\n",
    "from collections import Counter\n",
    "from sklearn.cluster import KMeans\n",
    "import matplotlib.pyplot as plt\n",
    "# from sklearn.metrics import confusion_matrix, ConfusionMatrixDisplay\n",
    "import sys\n",
    "import seaborn as sns\n",
    "# import pyclustering\n",
    "import matplotlib.pylab as plt\n",
    "import seaborn as sb\n",
    "from utilities.general_helpers import *\n",
    "from os.path import join\n",
    "from utilities.droplet_dataset import get_requested_subset\n",
    "from shutil import copyfile\n",
    "import matplotlib.pyplot as plt\n",
    "import torch\n",
    "import torch.nn as nn\n",
    "\n",
    "from torch.utils.data import Dataset\n",
    "from torchvision import datasets\n",
    "from torchvision.transforms import ToTensor\n",
    "from torch.utils.data import DataLoader\n",
    "from tqdm import tqdm"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Loads data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading cohort file:\n",
      "/storage/md_keren/shitay/Data/droplet_seq/cohort/normalized/6.21/cohort_normalized_26.6.21_protein_coding_genes.pkl\n"
     ]
    }
   ],
   "source": [
    "FILTERED_CELLS_PATH = r'/storage/md_keren/shitay/Data/droplet_seq/cohort/normalized/6.21/cohort_normalized_26.6.21_protein_coding_genes.pkl'\n",
    "print(f'Loading cohort file:\\n{FILTERED_CELLS_PATH}')\n",
    "cohort = pickle.load(open(FILTERED_CELLS_PATH, 'rb'))\n",
    "filtered_cells = cohort.filter_cells_by_property('is_immune', True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Classifier func:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Classifier(nn.Module):\n",
    "\n",
    "    def __init__(self, output_dim=None, input_size=None):\n",
    "        \"\"\"\n",
    "\n",
    "        The Classifer class: We are developing a model similar to ACTINN for good accuracy\n",
    "\n",
    "        \"\"\"\n",
    "        if output_dim == None or input_size == None:\n",
    "            raise ValueError('Must explicitly declare input dim (num features) and output dim (number of classes)')\n",
    "\n",
    "        super(Classifier, self).__init__();\n",
    "        self.inp_dim = input_size;\n",
    "        self.out_dim = output_dim;\n",
    "\n",
    "        # feed forward layers\n",
    "        self.classifier_sequential = nn.Sequential(\n",
    "                                        nn.Linear(self.inp_dim, 100),\n",
    "                                        nn.ReLU(),\n",
    "\n",
    "                                        nn.Linear(100, 50),\n",
    "                                        nn.ReLU(),\n",
    "\n",
    "                                        nn.Linear(50, 25),\n",
    "                                        nn.ReLU(),\n",
    "\n",
    "                                        nn.Linear(25, output_dim)\n",
    "                                        # SoftMax not needed for CrossEntropyLoss in PyTorch\n",
    "                                        ## i.e. no need for nn.Softmax(dim=1)\n",
    "                                        )\n",
    "\n",
    "    def forward(self, x):\n",
    "        \"\"\"\n",
    "\n",
    "        Forward pass of the classifier\n",
    "\n",
    "        \"\"\"\n",
    "\n",
    "        out = self.classifier_sequential(x);\n",
    "\n",
    "        return out"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Dataset object:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_response_labels():\n",
    "    # Loads xlsx files\n",
    "        CLINICAL_LABELS_PATH = r'/storage/md_keren/shitay/Data/tables/clinical_labels.xlsx'\n",
    "        MELANOMA_CLINICAL_DATA_PATH = r'/storage/md_keren/shitay/Data/tables/edited_unprotected_Melanoma_clinical_data.xlsx'\n",
    "        melanoma_clinical_data = pd.read_excel(MELANOMA_CLINICAL_DATA_PATH)\n",
    "        clinical_labels = pd.read_excel(CLINICAL_LABELS_PATH)\n",
    "\n",
    "        # takes nly first 46 samples, fill Nan and creat dictionary mapping\n",
    "        melanoma_clinical_data = melanoma_clinical_data.iloc[:46,:][['Patient id', 'Clinical response']]\n",
    "        melanoma_clinical_data['Clinical response'] = melanoma_clinical_data['Clinical response'].fillna('??')\n",
    "        labels_mapping = {v[0]: v[1] for v in clinical_labels[['Clinical response', 'binary label']].values}\n",
    "\n",
    "        # adds CRITERIA response into clinical df\n",
    "        melanoma_clinical_data['response'] = melanoma_clinical_data['Clinical response'].apply(lambda x: labels_mapping[x])\n",
    "        res_melanoma_clinical_data = melanoma_clinical_data.set_index('Patient id')\n",
    "        res_melanoma_clinical_data = [res_melanoma_clinical_data.loc[sample]['response'] for sample in cohort.samples]\n",
    "        return [0 if xx=='NR' else 1 for xx in res_melanoma_clinical_data]\n",
    "r_labels =  get_response_labels()  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "GENES = 20000\n",
    "CELLS = 79000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "class SCDataset(Dataset):\n",
    "    def __init__(self, cohort, transform=None, target_transform=None):       \n",
    "        self.X = torch.Tensor(cohort.counts[:CELLS, :GENES])\n",
    "        self.Y = r_labels[:CELLS]\n",
    "        self.transform = transform\n",
    "        self.target_transform = target_transform\n",
    "\n",
    "    def __len__(self):\n",
    "        return len(self.Y)\n",
    "\n",
    "    def __getitem__(self, idx):\n",
    "        X = self.X[idx]\n",
    "        label = self.Y[idx]\n",
    "        if self.transform:\n",
    "            X = self.transform(X)\n",
    "        if self.target_transform:\n",
    "            label = self.target_transform(label)\n",
    "        return X, label"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset = SCDataset(filtered_cells)\n",
    "\n",
    "train_dataloader = DataLoader(dataset, batch_size=64, shuffle=True)\n",
    "test_dataloader = DataLoader(dataset, batch_size=64, shuffle=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Builds a classifier, Training optimization and loss function:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      " Classifier Model \n",
      "\n",
      "Classifier(\n",
      "  (classifier_sequential): Sequential(\n",
      "    (0): Linear(in_features=20000, out_features=100, bias=True)\n",
      "    (1): ReLU()\n",
      "    (2): Linear(in_features=100, out_features=50, bias=True)\n",
      "    (3): ReLU()\n",
      "    (4): Linear(in_features=50, out_features=25, bias=True)\n",
      "    (5): ReLU()\n",
      "    (6): Linear(in_features=25, out_features=2, bias=True)\n",
      "  )\n",
      ")\n"
     ]
    }
   ],
   "source": [
    "classifier = Classifier(output_dim=2, input_size=GENES)\n",
    "\n",
    "cf_criterion = torch.nn.CrossEntropyLoss()\n",
    "\n",
    "#     cf_criterion = torch.nn.BCEWithLogitsLoss()\n",
    "cf_optimizer = torch.optim.Adam(params=classifier.parameters(), \n",
    "                                lr=0.0001, \n",
    "                                betas=(0.9, 0.999), \n",
    "                                eps=1e-08, \n",
    "                                weight_decay=0.005, \n",
    "                                amsgrad=False)\n",
    "cf_decayRate = 0.95\n",
    "cf_lr_scheduler = torch.optim.lr_scheduler.ExponentialLR(optimizer=cf_optimizer, gamma=cf_decayRate)\n",
    "print(\"\\n Classifier Model \\n\")\n",
    "print(classifier)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Trains:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Accuracy before train: 0.7481139240506329\n"
     ]
    }
   ],
   "source": [
    "def accuracy():\n",
    "    all_data_x = dataset[:][0]\n",
    "    all_data_y = dataset[:][1]\n",
    "    # all_data_y\n",
    "    res = classifier(all_data_x).detach().numpy().argmax(1) == np.array(all_data_y)\n",
    "    return sum(res)/len(res)\n",
    "\n",
    "print(f'Accuracy before train: {accuracy()}')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Classifier Training:   0%|          | 0/11 [00:00<?, ?it/s]/Local/md_keren/anaconda3/lib/python3.8/site-packages/torch/autograd/__init__.py:130: UserWarning: CUDA initialization: Found no NVIDIA driver on your system. Please check that you have an NVIDIA GPU and installed a driver from http://www.nvidia.com/Download/index.aspx (Triggered internally at  /pytorch/c10/cuda/CUDAFunctions.cpp:100.)\n",
      "  Variable._execution_engine.run_backward(\n",
      "Classifier Training: 100%|██████████| 11/11 [10:05<00:00, 55.00s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Accuracy after train: 0.993126582278481\n"
     ]
    }
   ],
   "source": [
    "ClassifierEpochs = 10\n",
    "\n",
    "\n",
    "def train_loop(dataloader, model, loss_fn, optimizer):\n",
    "    size = len(dataloader.dataset)\n",
    "    for batch, (X, y) in enumerate(dataloader):\n",
    "        # Compute prediction and loss\n",
    "#         print(X.shape)\n",
    "#         print(X)\n",
    "#         print(y)\n",
    "        \n",
    "        pred = model(X)\n",
    "        loss = loss_fn(pred, y.type(torch.long))\n",
    "\n",
    "        # Backpropagation\n",
    "        optimizer.zero_grad()\n",
    "        loss.backward()\n",
    "        optimizer.step()\n",
    "\n",
    "\n",
    "for epoch in tqdm(range(0, ClassifierEpochs + 1), desc=\"Classifier Training\"): \n",
    "\n",
    "    classifier.train()\n",
    "    train_loop(train_dataloader, classifier, cf_criterion, cf_optimizer) \n",
    "    \n",
    "print(f'Accuracy after train: {accuracy()}')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# all_data_x = dataset[:][0]\n",
    "# all_data_y = dataset[:][1]\n",
    "# # all_data_y\n",
    "# res = classifier(all_data_x).detach().numpy().argmax(1) == np.array(all_data_y)\n",
    "# dataset[:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Debug"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 502,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "tensor(0.0001)"
      ]
     },
     "execution_count": 502,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cf_criterion(torch.Tensor([[.1,10.12,.7]]), \n",
    "             torch.Tensor([1]).type(torch.long))\n",
    "# cf_criterion\n",
    "# torch.Tensor([[1]]).type(torch.int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 319,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "tensor(2.3610, grad_fn=<NllLossBackward>)"
      ]
     },
     "execution_count": 319,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loss = nn.CrossEntropyLoss()\n",
    "input = torch.randn(3, 5, requires_grad=True)\n",
    "target = torch.empty(3, dtype=torch.long).random_(5)\n",
    "loss(input, target)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 497,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "tensor([[ 7.9772, -7.9510],\n",
       "        [ 4.8489, -4.9263],\n",
       "        [ 9.2736, -9.2079]], grad_fn=<AddmmBackward>)"
      ]
     },
     "execution_count": 497,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input.dtype\n",
    "# target.dtype\n",
    "classifier(torch.randn(3, 20000, requires_grad=True))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # # Display image and label.\n",
    "# train_features, train_labels = next(iter(train_dataloader))\n",
    "# print(f\"Feature batch shape: {train_features}\")\n",
    "# print(f\"Labels batch shape: {train_labels}\")\n",
    "# # img = train_features[0]\n",
    "# # label = train_labels[0]\n",
    "# # plt.imshow(img, cmap=\"gray\")\n",
    "# # plt.show()\n",
    "# # print(f\"Label: {label}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 507,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Loads xlsx files\n",
    "CLINICAL_LABELS_PATH = r'/storage/md_keren/shitay/Data/tables/clinical_labels.xlsx'\n",
    "MELANOMA_CLINICAL_DATA_PATH = r'/storage/md_keren/shitay/Data/tables/edited_unprotected_Melanoma_clinical_data.xlsx'\n",
    "melanoma_clinical_data = pd.read_excel(MELANOMA_CLINICAL_DATA_PATH)\n",
    "clinical_labels = pd.read_excel(CLINICAL_LABELS_PATH)\n",
    "\n",
    "# takes nly first 46 samples, fill Nan and creat dictionary mapping\n",
    "melanoma_clinical_data = melanoma_clinical_data.iloc[:46,:][['Patient id', 'Clinical response']]\n",
    "melanoma_clinical_data['Clinical response'] = melanoma_clinical_data['Clinical response'].fillna('??')\n",
    "labels_mapping = {v[0]: v[1] for v in clinical_labels[['Clinical response', 'binary label']].values}\n",
    "\n",
    "# adds CRITERIA response into clinical df\n",
    "melanoma_clinical_data['response'] = melanoma_clinical_data['Clinical response'].apply(lambda x: labels_mapping[x])\n",
    "res_melanoma_clinical_data = melanoma_clinical_data.set_index('Patient id')\n",
    "# res_melanoma_clinical_data = [res_melanoma_clinical_data.loc[sample]['response'] for sample in cohort.samples]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 511,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#</th>\n",
       "      <th>Patient id</th>\n",
       "      <th>MGH id</th>\n",
       "      <th>Date collected</th>\n",
       "      <th>Sex (F/M)</th>\n",
       "      <th>Age (relevant to time of sample)</th>\n",
       "      <th>Melanoma type</th>\n",
       "      <th>Primary=1; Metastasis=0</th>\n",
       "      <th>Genotype</th>\n",
       "      <th>Therapy(ies) prior to biopsy</th>\n",
       "      <th>Therapy after biopsy</th>\n",
       "      <th>Clinical response</th>\n",
       "      <th>Status (Alive=1/Dead=0)</th>\n",
       "      <th>Biopsy site</th>\n",
       "      <th>Biopsy type (pre, post)</th>\n",
       "      <th>Overall survival (days)</th>\n",
       "      <th>Comments</th>\n",
       "      <th>Cell line (Y; N)</th>\n",
       "      <th>TILs (Y; N)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>M97</td>\n",
       "      <td>1256</td>\n",
       "      <td>2019-10-11</td>\n",
       "      <td>M</td>\n",
       "      <td>79.0</td>\n",
       "      <td>Cutaneous</td>\n",
       "      <td>0</td>\n",
       "      <td>TP63; DDR2; ATM; MAP2K1; HNF1A; TERT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PD1 (started on 3/31/20)</td>\n",
       "      <td>R</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Left Axillary LN</td>\n",
       "      <td>Pre</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>M98</td>\n",
       "      <td>509</td>\n",
       "      <td>2019-10-11</td>\n",
       "      <td>M</td>\n",
       "      <td>50.0</td>\n",
       "      <td>Cutaneous</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>aPD1 (progressed on 7/29/19)- 18 cycles, start...</td>\n",
       "      <td>NA (surgical management)</td>\n",
       "      <td>R (had a response in 2016-2019)</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Right inguinal LN- progressing lesion</td>\n",
       "      <td>Post</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1126 days- PFS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>M99</td>\n",
       "      <td>1063</td>\n",
       "      <td>2019-10-11</td>\n",
       "      <td>F</td>\n",
       "      <td>69.0</td>\n",
       "      <td>Mucosal</td>\n",
       "      <td>0</td>\n",
       "      <td>NF2, NRAS, CTNNB1 variants</td>\n",
       "      <td>5/11/17- ipi+nivo; 7/18/17- nivo; 1/8/18- pemb...</td>\n",
       "      <td>TVEC</td>\n",
       "      <td>NR (PD)</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Left distal femur curettage</td>\n",
       "      <td>Post</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>M100</td>\n",
       "      <td>592</td>\n",
       "      <td>2019-10-21</td>\n",
       "      <td>M</td>\n",
       "      <td>88.0</td>\n",
       "      <td>Cutaneous</td>\n",
       "      <td>0</td>\n",
       "      <td>NRAS, TERT, RET, NF1, SMARCA4 variants</td>\n",
       "      <td>1/3/19- pembro (one dose-toxicity)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NR (SD)</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Endoscopy</td>\n",
       "      <td>Post</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>M101</td>\n",
       "      <td>1259</td>\n",
       "      <td>2019-10-21</td>\n",
       "      <td>M</td>\n",
       "      <td>57.0</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>0</td>\n",
       "      <td>TRET, TP53, NRAS variants</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pembro (adj)- 1/14/20</td>\n",
       "      <td>NED- 10/5/20</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Left inguinal LN (first met diagnosis)</td>\n",
       "      <td>Pre adj</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>87</td>\n",
       "      <td>M190</td>\n",
       "      <td>1608</td>\n",
       "      <td>2021-03-11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>88</td>\n",
       "      <td>M191</td>\n",
       "      <td>1571</td>\n",
       "      <td>2021-03-22</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>89</td>\n",
       "      <td>M192</td>\n",
       "      <td>1487</td>\n",
       "      <td>2021-03-25</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>90</td>\n",
       "      <td>M193</td>\n",
       "      <td>1622</td>\n",
       "      <td>2021-03-26</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>91</td>\n",
       "      <td>M194</td>\n",
       "      <td>896</td>\n",
       "      <td>2021-04-05</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>91 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     # Patient id MGH id Date collected Sex (F/M)  \\\n",
       "0    1        M97   1256     2019-10-11         M   \n",
       "1    2        M98    509     2019-10-11         M   \n",
       "2    3        M99   1063     2019-10-11         F   \n",
       "3    4       M100    592     2019-10-21         M   \n",
       "4    5       M101   1259     2019-10-21         M   \n",
       "..  ..        ...    ...            ...       ...   \n",
       "86  87       M190   1608     2021-03-11       NaN   \n",
       "87  88       M191   1571     2021-03-22       NaN   \n",
       "88  89       M192   1487     2021-03-25       NaN   \n",
       "89  90       M193   1622     2021-03-26       NaN   \n",
       "90  91       M194    896     2021-04-05       NaN   \n",
       "\n",
       "    Age (relevant to time of sample) Melanoma type Primary=1; Metastasis=0  \\\n",
       "0                               79.0     Cutaneous                       0   \n",
       "1                               50.0     Cutaneous                       0   \n",
       "2                               69.0      Mucosal                        0   \n",
       "3                               88.0     Cutaneous                       0   \n",
       "4                               57.0       Unknown                       0   \n",
       "..                               ...           ...                     ...   \n",
       "86                               NaN           NaN                     NaN   \n",
       "87                               NaN           NaN                     NaN   \n",
       "88                               NaN           NaN                     NaN   \n",
       "89                               NaN           NaN                     NaN   \n",
       "90                               NaN           NaN                     NaN   \n",
       "\n",
       "                                 Genotype   \\\n",
       "0     TP63; DDR2; ATM; MAP2K1; HNF1A; TERT   \n",
       "1                                      NaN   \n",
       "2               NF2, NRAS, CTNNB1 variants   \n",
       "3   NRAS, TERT, RET, NF1, SMARCA4 variants   \n",
       "4                TRET, TP53, NRAS variants   \n",
       "..                                     ...   \n",
       "86                                     NaN   \n",
       "87                                     NaN   \n",
       "88                                     NaN   \n",
       "89                                     NaN   \n",
       "90                                     NaN   \n",
       "\n",
       "                         Therapy(ies) prior to biopsy  \\\n",
       "0                                                 NaN   \n",
       "1   aPD1 (progressed on 7/29/19)- 18 cycles, start...   \n",
       "2   5/11/17- ipi+nivo; 7/18/17- nivo; 1/8/18- pemb...   \n",
       "3                 1/3/19- pembro (one dose-toxicity)    \n",
       "4                                                 NaN   \n",
       "..                                                ...   \n",
       "86                                                NaN   \n",
       "87                                                NaN   \n",
       "88                                                NaN   \n",
       "89                                                NaN   \n",
       "90                                                NaN   \n",
       "\n",
       "        Therapy after biopsy                Clinical response  \\\n",
       "0   PD1 (started on 3/31/20)                                R   \n",
       "1   NA (surgical management)  R (had a response in 2016-2019)   \n",
       "2                       TVEC                          NR (PD)   \n",
       "3                        NaN                          NR (SD)   \n",
       "4      Pembro (adj)- 1/14/20                     NED- 10/5/20   \n",
       "..                       ...                              ...   \n",
       "86                       NaN                              NaN   \n",
       "87                       NaN                              NaN   \n",
       "88                       NaN                              NaN   \n",
       "89                       NaN                              NaN   \n",
       "90                       NaN                              NaN   \n",
       "\n",
       "    Status (Alive=1/Dead=0)                             Biopsy site  \\\n",
       "0                       1.0                        Left Axillary LN   \n",
       "1                       1.0   Right inguinal LN- progressing lesion   \n",
       "2                       1.0             Left distal femur curettage   \n",
       "3                       1.0                               Endoscopy   \n",
       "4                       1.0  Left inguinal LN (first met diagnosis)   \n",
       "..                      ...                                     ...   \n",
       "86                      NaN                                     NaN   \n",
       "87                      NaN                                     NaN   \n",
       "88                      NaN                                     NaN   \n",
       "89                      NaN                                     NaN   \n",
       "90                      NaN                                     NaN   \n",
       "\n",
       "   Biopsy type (pre, post) Overall survival (days)        Comments  \\\n",
       "0                      Pre                     NaN             NaN   \n",
       "1                     Post                     NaN  1126 days- PFS   \n",
       "2                     Post                     NaN             NaN   \n",
       "3                     Post                     NaN             NaN   \n",
       "4                  Pre adj                     NaN             NaN   \n",
       "..                     ...                     ...             ...   \n",
       "86                     NaN                     NaN             NaN   \n",
       "87                     NaN                     NaN             NaN   \n",
       "88                     NaN                     NaN             NaN   \n",
       "89                     NaN                     NaN             NaN   \n",
       "90                     NaN                     NaN             NaN   \n",
       "\n",
       "   Cell line (Y; N) TILs (Y; N)  \n",
       "0               NaN         NaN  \n",
       "1               NaN         NaN  \n",
       "2               NaN         NaN  \n",
       "3               NaN         NaN  \n",
       "4               NaN         NaN  \n",
       "..              ...         ...  \n",
       "86              NaN         NaN  \n",
       "87              NaN         NaN  \n",
       "88              NaN         NaN  \n",
       "89              NaN         NaN  \n",
       "90              NaN         NaN  \n",
       "\n",
       "[91 rows x 19 columns]"
      ]
     },
     "execution_count": 511,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_excel(MELANOMA_CLINICAL_DATA_PATH)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
