{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "lib = r'/srv01/technion/shitay/Code/classifying_response_to_immunotherapy/'\n",
    "import sys\n",
    "sys.path.append(lib)\n",
    "from utilities.package_importing import *\n",
    "from statsmodels.stats.weightstats import ztest\n",
    "import scanpy as sc\n",
    "import anndata\n",
    "import warnings \n",
    "warnings.filterwarnings(\"ignore\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "### HELPER FUNCTIONS ###\n",
    "\n",
    "def fraction_test(df_expression, FEATURE, gene, cell_type):\n",
    "    # return pval\n",
    "    df_expression['is_expressing'] = df_expression.exp >2.5\n",
    "    values_in_feature = list(set(df_expression[FEATURE]))\n",
    "    if len(values_in_feature) < 2:\n",
    "        print(f'no cells in one of the groups for: {gene}, {cell_type}')\n",
    "        return 1\n",
    "    val_1 = values_in_feature[0]\n",
    "    val_2 = values_in_feature[1]\n",
    "    \n",
    "    if len(df_expression[df_expression[FEATURE]==val_1])<=5 or len(df_expression[df_expression[FEATURE]==val_2])<=5:\n",
    "        return 1\n",
    "    fractions = df_expression.groupby(['sample', FEATURE])['is_expressing'].agg(['sum', 'count']).reset_index()\n",
    "    fractions['fractions'] = fractions['sum']/fractions['count']\n",
    "    statistic, pvalue = ranksums(fractions[fractions[FEATURE]==val_2].fractions, fractions[fractions[FEATURE]==val_1].fractions)\n",
    "    return pvalue\n",
    "\n",
    "def Z_TEST(df_expression, FEATURE, gene, cell_type):\n",
    "    # return pval\n",
    "    values_in_feature = list(set(df_expression[FEATURE]))\n",
    "    if len(values_in_feature) < 2:\n",
    "        print(f'no cells in one of the groups for: {gene}, {cell_type}')\n",
    "        return 1\n",
    "    val_1 = values_in_feature[0]\n",
    "    val_2 = values_in_feature[1]\n",
    "    \n",
    "    if len(df_expression[df_expression[FEATURE]==val_1])<=5 or len(df_expression[df_expression[FEATURE]==val_2])<=5:\n",
    "        return 1\n",
    "    statistic, pvalue = ztest(res_df[df_expression[FEATURE]==val_1].exp, df_expression[res_df[FEATURE]==val_2].exp, value=0) \n",
    "    return pvalue\n",
    "\n",
    "def fisher_test(df_expression, FEATURE, gene, cell_type):\n",
    "\n",
    "    df_expression['is_expressing'] = df_expression.exp >2.5\n",
    "    values_in_feature = list(set(df_expression[FEATURE]))\n",
    "    if len(values_in_feature) < 2:\n",
    "        print(f'no cells in one of the groups for: {gene}, {cell_type}')\n",
    "        return 1\n",
    "    val_1 = values_in_feature[0]\n",
    "    val_2 = values_in_feature[1]\n",
    "\n",
    "    \n",
    "    if len(df_expression[df_expression[FEATURE]==val_1])<=5 or len(df_expression[df_expression[FEATURE]==val_2])<=5:\n",
    "        return 1\n",
    "\n",
    "    n_cells_expressing = df_expression.groupby(FEATURE)['is_expressing'].sum()\n",
    "    n_cells = df_expression.groupby(FEATURE)['is_expressing'].count()\n",
    "\n",
    "    R_n_cells_expressing = n_cells_expressing[val_1]\n",
    "    NR_n_cells_expressing = n_cells_expressing[val_2]\n",
    "    R_n_cells_not_expressing = n_cells[val_1] - R_n_cells_expressing\n",
    "    NR_n_cells_not_expressing = n_cells[val_2] - NR_n_cells_expressing\n",
    "\n",
    "    import scipy.stats as stats\n",
    "    oddsratio, pvalue = stats.fisher_exact([[R_n_cells_expressing, NR_n_cells_expressing],\n",
    "                                            [R_n_cells_not_expressing, NR_n_cells_not_expressing]])\n",
    "    return pvalue\n",
    "\n",
    "# The rule is * for p<0.05, ** for p<0.005 and *** for p<0.0005.\n",
    "\n",
    "def get_pval_symbol(p):\n",
    "    if p<0.0005:\n",
    "        return '***'\n",
    "    elif p<0.005:\n",
    "        return '**'\n",
    "    elif p<0.05:\n",
    "        return '*'\n",
    "    else:\n",
    "        return ''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path = r'/storage/md_keren/shitay/Data/smart_seq/SmartSeq_RNAseq_DATA.p'\n",
    "cells, gene_names, patients_information = extract_smart_seq_data_from_pickle(data_path)\n",
    "\n",
    "from utilities.smart_seq_dataset import RNAseq_Dataset\n",
    "cohort = RNAseq_Dataset(cells, patients_information, gene_names)\n",
    "\n",
    "### Prob: TNFA appears as TNF\n",
    "\n",
    "genes = ['CCR8',  'CXCR3', 'CCL1', 'CXCL11', 'CXCL10', 'CXCL9', 'IFNG', 'TNF', 'GZMB', 'PRF1', 'IL2']\n",
    "gene_indices = [cohort.gene_names.index(gg) for gg in genes]\n",
    "\n",
    "from anndata import AnnData\n",
    "annCohort = AnnData(cohort.cells)\n",
    "\n",
    "annCohort.var['gene'] = cohort.gene_names\n",
    "\n",
    "annCohort.obs['sample'] = [cell.patient_details for cell in cohort.cells_information_list] \n",
    "annCohort.obs['barcode'] = [cell.cell_id for cell in cohort.cells_information_list] \n",
    "annCohort.obs['cell_type'] = [cell.supervised for cell in cohort.cells_information_list] \n",
    "annCohort.obs['response'] = [cell.response for cell in cohort.cells_information_list] \n",
    "annCohort.obs['pre/post'] = [p.split('_')[0] for p in annCohort.obs['sample'].values.tolist()]\n",
    "\n",
    "### Mark cells as CD103+ is they express CD103\n",
    "MIN_EXP_THRESHOLD = 2.5\n",
    "CD103_index = annCohort.var.gene.tolist().index('ITGAE') #.index('CD103')\n",
    "annCohort.obs['CD103+'] = annCohort.X[:,CD103_index] > MIN_EXP_THRESHOLD\n",
    "\n",
    "### Mark cells as myeloid if their cell type is myeloid\n",
    "MYELOID_CELL_TYPES = ['Macrophage_immature', 'Macrophage_mature', 'Monocyte_immature', 'Monocyte_mature', 'cDCs_dendritic_cells', \n",
    " 'pDCs', 'myeloid cells_general_immature', 'myeloid cells_general_mature', 'Neutrophils', 'Granolocytes']\n",
    "def is_overlap(l1 , l2):\n",
    "    return len([f for f in l1 if f in l2])!=0\n",
    "annCohort.obs['is_myeloid'] = annCohort.obs['cell_type'].apply(lambda x: is_overlap(x, MYELOID_CELL_TYPES))\n",
    "\n",
    "\n",
    "### Update cell type list\n",
    "def add_pos_CD103_in_df(x):\n",
    "    if x['CD103+'] and x['is_myeloid']:\n",
    "        return x['cell_type'] + ['CD103+']\n",
    "    else:\n",
    "        return x['cell_type'] \n",
    "\n",
    "annCohort.obs['cell_type'] = annCohort.obs.apply(lambda x: add_pos_CD103_in_df(x), axis=1)\n",
    "\n",
    "cell_types = list(set(flatten_list(annCohort.obs['cell_type'].tolist())))\n",
    "t_cell_types = [ct for ct in cell_types if 'T' in ct]\n",
    "not_t_cell_types = [ct for ct in cell_types if not ct in t_cell_types]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. More customized run"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "CELL_TYPES = ['T cells', 'Activated T cells', 'CD8 Cytotoxic T cells', 'Regulatory T cells', 'CD4 helper T cells']\n",
    "genes = ['CXCR3', 'CCL1'] #, 'CXCL11', 'CXCL10', 'CXCL9', 'IFNG', 'TNF', 'GZMB', 'PRF1', 'IL2']\n",
    "FEATURE = 'pre/post'\n",
    "filter_type = 1\n",
    "\n",
    "fig_size_inches = (10, 6)\n",
    "font_size = 16"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gonna filter response\n",
      "took only Responder samples\n",
      "cohort for comparison details:\n",
      "View of AnnData object with n_obs × n_vars = 5564 × 20214\n",
      "    obs: 'sample', 'barcode', 'cell_type', 'response', 'pre/post', 'CD103+', 'is_myeloid'\n",
      "    var: 'gene'\n"
     ]
    }
   ],
   "source": [
    "### builds DFs of all genes X cell types and conduct Z test:\n",
    "# fig_size_inches = (32, 30)\n",
    "# font_size = 32\n",
    "\n",
    "################################## TEST_NAME ##################################\n",
    "TEST_NAME='FISHER'\n",
    "stat_test = fisher_test\n",
    "################################## Filter type ##################################\n",
    "filter_dic = {'response': {1: 'Responder', 2: 'Non-responder'}, 'pre/post': {1: 'Pre', 2:'Post'}}\n",
    "filter_feature_dic = {'response': 'pre/post', 'pre/post': 'response'} # take by reverse\n",
    "\n",
    "filtered_cohort = annCohort\n",
    "if filter_type:\n",
    "    FILTER_FEATURE = filter_feature_dic[FEATURE]\n",
    "    print(f'gonna filter {FILTER_FEATURE}')\n",
    "    filter_feature_req_value = filter_dic[FILTER_FEATURE][filter_type]\n",
    "    filtered_cohort = annCohort[annCohort.obs[FILTER_FEATURE] == filter_feature_req_value]\n",
    "    print(f'took only {filter_feature_req_value} samples')\n",
    "\n",
    "else:\n",
    "    print(f'did not filter')\n",
    "print(f'cohort for comparison details:')\n",
    "print(filtered_cohort)\n",
    "\n",
    "\n",
    "################################## TO_TEST_T_CELLS ##################################\n",
    "df_fractions = pd.DataFrame(columns=['sample', 'response', 'fraction', 'gene', 'cell_type'])\n",
    "pvals = {}\n",
    "for gene_idx, gene in enumerate(genes):\n",
    "    gene_cohort = filtered_cohort[:,  filtered_cohort.var['gene']==gene]\n",
    "    for idx, cell_type in enumerate(CELL_TYPES):\n",
    "        cell_type_indices = [cell_type in ttt for ttt in gene_cohort.obs['cell_type']]\n",
    "        cell_type_cohort = gene_cohort[cell_type_indices, :]\n",
    "        \n",
    "        gene_exps = cell_type_cohort.X[:,0].tolist()\n",
    "        res_df = pd.DataFrame(np.concatenate((cell_type_cohort.obs[['sample', FEATURE]].values.T,[gene_exps])).T, columns=['sample', FEATURE, 'exp'])\n",
    "        \n",
    "        ### statistical test of gene expression R/NR\n",
    "        pvalue = stat_test(res_df, FEATURE, gene, cell_type)\n",
    "        pvals[(gene, cell_type)] = pvalue\n",
    "        \n",
    "        sample_fraction_df = pd.DataFrame(res_df.groupby(['sample', FEATURE]).sum()['is_expressing'] / res_df.groupby('sample').count()['is_expressing']).reset_index().rename(columns={'is_expressing': 'fraction'})\n",
    "        sample_fraction_df['gene'] = gene\n",
    "        sample_fraction_df['cell_type'] = cell_type\n",
    "        df_fractions = df_fractions.append(sample_fraction_df)\n",
    "\n",
    "# Correct pvals\n",
    "pvals_arr = np.array([[k, v] for k,v in list(pvals.items())])\n",
    "corrected_pvals = pvals_arr.copy()\n",
    "corrected_pvals[:,1] = multipletests_fdr(pvals_arr[:,1])[1].tolist()\n",
    "corrected_pvals = {pp[0]: pp[1] for pp in corrected_pvals}\n",
    "\n",
    "# Prints using sns\n",
    "fig, axs = plt.subplots(len(genes))\n",
    "fig.set_size_inches(fig_size_inches)\n",
    "def get_coorected_label(x ,p_value_dic):\n",
    "    pvalue = p_value_dic[(x['gene'], x['cell_type'])]\n",
    "    return x['cell_type'] + '\\n'+''.join(['*' for sig in [0.05, 0.005, 0.0005] if pvalue < sig])\n",
    "df_fractions['labeled_cell_type'] = df_fractions.apply(lambda x: get_coorected_label(x, corrected_pvals), axis=1)\n",
    "\n",
    "hue_order = {'pre/post': [ \"Pre\", \"Post\"],\n",
    "             'response': [\"Non-responder\", \"Responder\"]}\n",
    "for gene_idx, gene in enumerate(genes):\n",
    "        df_fractions_cur_gene = df_fractions[df_fractions['gene'] == gene]\n",
    "        sns.boxplot(ax=axs[gene_idx], x='labeled_cell_type', y='fraction', hue=FEATURE, hue_order = hue_order[FEATURE],\n",
    "                    data=df_fractions_cur_gene, palette={\"Responder\": \"b\", \"Non-responder\": \"r\", \"Pre\":\"r\", \"Post\":\"b\"}).set(title=gene);\n",
    "        sns.stripplot(ax=axs[gene_idx], x='labeled_cell_type', y='fraction', data=df_fractions_cur_gene, hue=FEATURE, hue_order = hue_order[FEATURE],\n",
    "                  jitter=True, dodge=True, alpha=0.3,linewidth=0.5, palette={\"Responder\": \"b\", \"Non-responder\": \"r\", \"Pre\":\"r\", \"Post\":\"b\"});\n",
    "        if not (gene_idx==0): axs[gene_idx].legend_ = None\n",
    "\n",
    "if filter_type:\n",
    "    fig.suptitle(f'Comparing {FEATURE}, only \\'{filter_feature_req_value}\\' samples were taken', fontsize=font_size)\n",
    "else:\n",
    "    fig.suptitle(f'Comparing {FEATURE}', fontsize=font_size)\n",
    "fig.tight_layout(rect=[0, 0.03, 1, 0.95])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Only set output dir and all combination of experiments will be run and be saved there:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "OUTPUT_FOLDER = r'/storage/md_keren/shitay/Nathi_project/differential analysis/20.12.21'\n",
    "\n",
    "combinations = [[c1, c2, c3, c4] for c1 in ['FISHER', 'FRACTION'] for c2 in [\"response\", \"pre/post\"] for c3 in [None, 1,2] for c4 in [True, False]]\n",
    "create_folder(OUTPUT_FOLDER)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for c_idx, comb in enumerate(combinations):\n",
    "    print(f'\\n######################################################################################################')\n",
    "    print(f'Working on {comb} ({c_idx+1}/{len(combinations)})')\n",
    "    TEST_NAME = comb[0] # FRACTION/Z/FISHER\n",
    "    FEATURE = comb[1] # \"response\" or \"pre/post\"\n",
    "    filter_type = comb[2]\n",
    "    TO_TEST_T_CELLS = comb[3] # True - t cells \\ False - not t cells - all other\n",
    "    ################################## TEST_NAME ##################################\n",
    "    if TEST_NAME=='FRACTION':\n",
    "        stat_test = fraction_test\n",
    "    elif TEST_NAME=='Z':\n",
    "        stat_test = Z_TEST\n",
    "    elif TEST_NAME=='FISHER':\n",
    "        stat_test = fisher_test\n",
    "    else:\n",
    "        raise Exception('statistical name is not valid, change it')\n",
    "\n",
    "\n",
    "    ################################## Filter type ##################################\n",
    "    filter_dic = {'response': {1: 'Responder', 2: 'Non-responder'}, 'pre/post': {1: 'Pre', 2:'Post'}}\n",
    "    filter_feature_dic = {'response': 'pre/post', 'pre/post': 'response'} # take by reverse\n",
    "\n",
    "    filtered_cohort = annCohort\n",
    "    if filter_type:\n",
    "        FILTER_FEATURE = filter_feature_dic[FEATURE]\n",
    "        print(f'gonna filter {FILTER_FEATURE}')\n",
    "        filter_feature_req_value = filter_dic[FILTER_FEATURE][filter_type]\n",
    "        filtered_cohort = annCohort[annCohort.obs[FILTER_FEATURE] == filter_feature_req_value]\n",
    "        print(f'took only {filter_feature_req_value} samples')\n",
    "\n",
    "    else:\n",
    "        print(f'did not filter')\n",
    "    print(f'cohort for comparison details:')\n",
    "    print(filtered_cohort)\n",
    "\n",
    "\n",
    "    ################################## TO_TEST_T_CELLS ##################################\n",
    "\n",
    "    if TO_TEST_T_CELLS:\n",
    "        CELL_TYPES = t_cell_types\n",
    "    else:\n",
    "        CELL_TYPES = not_t_cell_types\n",
    "\n",
    "    # Takes only cell types that appear in filted_cohort\n",
    "    CELL_TYPES = [ct for ct in CELL_TYPES if ct in list(set(flatten_list(filtered_cohort.obs['cell_type'].tolist())))]\n",
    "\n",
    "    ################################## Run experiment ##################################\n",
    "    df_fractions = pd.DataFrame(columns=['sample', 'response', 'fraction', 'gene', 'cell_type'])\n",
    "    pvals = {}\n",
    "    for gene_idx, gene in enumerate(genes):\n",
    "        gene_cohort = filtered_cohort[:,  filtered_cohort.var['gene']==gene]\n",
    "        for idx, cell_type in enumerate(CELL_TYPES):\n",
    "            cell_type_indices = [cell_type in ttt for ttt in gene_cohort.obs['cell_type']]\n",
    "            cell_type_cohort = gene_cohort[cell_type_indices, :]\n",
    "\n",
    "            gene_exps = cell_type_cohort.X[:,0].tolist()\n",
    "            res_df = pd.DataFrame(np.concatenate((cell_type_cohort.obs[['sample', FEATURE]].values.T,[gene_exps])).T, columns=['sample', FEATURE, 'exp'])\n",
    "\n",
    "            ### statistical test of gene expression R/NR\n",
    "            pvalue = stat_test(res_df, FEATURE, gene, cell_type)\n",
    "            pvals[(gene, cell_type)] = pvalue\n",
    "\n",
    "            sample_fraction_df = pd.DataFrame(res_df.groupby(['sample', FEATURE]).sum()['is_expressing'] / res_df.groupby('sample').count()['is_expressing']).reset_index().rename(columns={'is_expressing': 'fraction'})\n",
    "            sample_fraction_df['gene'] = gene\n",
    "            sample_fraction_df['cell_type'] = cell_type\n",
    "            df_fractions = df_fractions.append(sample_fraction_df)\n",
    "\n",
    "    # Correct pvals\n",
    "    pvals_arr = np.array([[k, v] for k,v in list(pvals.items())])\n",
    "    corrected_pvals = pvals_arr.copy()\n",
    "    corrected_pvals[:,1] = multipletests_fdr(pvals_arr[:,1])[1].tolist()\n",
    "    corrected_pvals = {pp[0]: pp[1] for pp in corrected_pvals}\n",
    "\n",
    "    # Prints using sns\n",
    "    fig, axs = plt.subplots(len(genes))\n",
    "    fig.set_size_inches(32, 30)\n",
    "    def get_coorected_label(x ,p_value_dic):\n",
    "        pvalue = p_value_dic[(x['gene'], x['cell_type'])]\n",
    "        return x['cell_type'] + '\\n'+''.join(['*' for sig in [0.05, 0.005, 0.0005] if pvalue < sig])\n",
    "    df_fractions['labeled_cell_type'] = df_fractions.apply(lambda x: get_coorected_label(x, corrected_pvals), axis=1)\n",
    "\n",
    "    hue_order = {'pre/post': [ \"Pre\", \"Post\"],\n",
    "                 'response': [\"Non-responder\", \"Responder\"]}\n",
    "    for gene_idx, gene in enumerate(genes):\n",
    "            df_fractions_cur_gene = df_fractions[df_fractions['gene'] == gene]\n",
    "            sns.boxplot(ax=axs[gene_idx], x='labeled_cell_type', y='fraction', hue=FEATURE, hue_order = hue_order[FEATURE],\n",
    "                        data=df_fractions_cur_gene, palette={\"Responder\": \"b\", \"Non-responder\": \"r\", \"Pre\":\"r\", \"Post\":\"b\"}).set(title=gene);\n",
    "            sns.stripplot(ax=axs[gene_idx], x='labeled_cell_type', y='fraction', data=df_fractions_cur_gene, hue=FEATURE, hue_order = hue_order[FEATURE],\n",
    "                      jitter=True, dodge=True, alpha=0.3,linewidth=0.5, palette={\"Responder\": \"b\", \"Non-responder\": \"r\", \"Pre\":\"r\", \"Post\":\"b\"});\n",
    "\n",
    "    if filter_type:\n",
    "        fig.suptitle(f'Comparing {FEATURE}, only \\'{filter_feature_req_value}\\' samples were taken, statistical_test={TEST_NAME}', fontsize=32)\n",
    "    else:\n",
    "        fig.suptitle(f'Comparing {FEATURE}, statistical_test={TEST_NAME}', fontsize=32)\n",
    "    fig.tight_layout(rect=[0, 0.03, 1, 0.95])\n",
    "\n",
    "\n",
    "\n",
    "    if TO_TEST_T_CELLS:\n",
    "        p_folder = 'T_cells'\n",
    "    else: p_folder = 'other_cells'\n",
    "    create_folder(join(OUTPUT_FOLDER, p_folder))\n",
    "\n",
    "\n",
    "\n",
    "    ################################## Save fig output ##################################\n",
    "    # folder = FEATURE.replace('/', '_').replace('response', 'R_NR')\n",
    "    test_n = TEST_NAME.lower().replace('fraction','ranksum')\n",
    "    feature_converted = FEATURE.replace('/', '_vs_')\n",
    "    file_name = fr'{p_folder.lower()}_{feature_converted}_comparison'\n",
    "    if filter_type:\n",
    "        file_name += f'_{filter_feature_req_value}_samples'\n",
    "    file_name += f'_{test_n}_test.png'\n",
    "    FULL_OUTPUT_PATH = join(OUTPUT_FOLDER, p_folder, file_name)\n",
    "    print(f'Will be saved in {FULL_OUTPUT_PATH}')\n",
    "    fig.savefig(FULL_OUTPUT_PATH)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
